api/cmd/build/main.go

```diff
@@ -0,0 +1,268 @@
+	fs.StringVar(&flagId, "id", "latest", "build id")
+	fs.StringVar(&flagMethod, "method", "", "source method")
+	fs.StringVar(&flagPush, "push", "", "push to registry")
+	fs.StringVar(&flagRelease, "release", "", "release id to fork")
+	fs.StringVar(&flagUrl, "url", "", "source url")
+
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		fail(err)
+	}
+
+	if v := os.Getenv("BUILD_APP"); v != "" {
+		flagApp = v
+	}
+
+	if v := os.Getenv("BUILD_AUTH"); v != "" {
+		flagAuth = v
+	}
+
+	if v := os.Getenv("BUILD_CONFIG"); v != "" {
+		flagConfig = v
+	}
+
+	if v := os.Getenv("BUILD_ID"); v != "" {
+		flagId = v
+	}
+
+	if v := os.Getenv("BUILD_PUSH"); v != "" {
+		flagPush = v
+	}
+
+	if v := os.Getenv("BUILD_RELEASE"); v != "" {
+		flagRelease = v
+	}
+
+	if v := os.Getenv("BUILD_URL"); v != "" {
+		flagUrl = v
+	}
+
+	// fmt.Printf("flagApp = %+v\n", flagApp)
+	// fmt.Printf("flagAuth = %+v\n", flagAuth)
+	// fmt.Printf("flagConfig = %+v\n", flagConfig)
+	// fmt.Printf("flagId = %+v\n", flagId)
+	// fmt.Printf("flagMethod = %+v\n", flagMethod)
+	// fmt.Printf("flagPush = %+v\n", flagPush)
+	// fmt.Printf("flagUrl = %+v\n", flagUrl)
+
+	if err := execute(); err != nil {
+		fail(err)
+	}
+
+	if err := success(); err != nil {
+		fail(err)
+	}
+}
+
+func execute() error {
+	b, err := provider.FromEnv().BuildGet(flagApp, flagId)
+	if err != nil {
+		return err
+	}
+
+	currentBuild = b
+
+	dir, err := fetch()
+	if err != nil {
+		return err
+	}
+
+	// defer os.RemoveAll(dir)
+
+	data, err := ioutil.ReadFile(filepath.Join(dir, flagConfig))
+	if err != nil {
+		return err
+	}
+
+	currentBuild.Manifest = string(data)
+
+	if err := login(); err != nil {
+		return err
+	}
+
+	if err := build(dir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func fetch() (string, error) {
+	var s source.Source
+
+	switch flagMethod {
+	case "index":
+		s = &source.SourceIndex{flagUrl}
+	case "tgz":
+		s = &source.SourceTgz{flagUrl}
+	case "zip":
+		s = &source.SourceZip{flagUrl}
+	default:
+		return "", fmt.Errorf("unknown method: %s", flagMethod)
+	}
+
+	return s.Fetch()
+}
+
+func login() error {
+	var auth map[string]struct {
+		Username string
+		Password string
+	}
+
+	if err := json.Unmarshal([]byte(flagAuth), &auth); err != nil {
+		return err
+	}
+
+	for host, entry := range auth {
+		out, err := exec.Command("docker", "login", "-u", entry.Username, "-p", entry.Password, host).CombinedOutput()
+		log(fmt.Sprintf("Authenticating %s: %s", host, strings.TrimSpace(string(out))))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func build(dir string) error {
+	dcy := filepath.Join(dir, flagConfig)
+
+	if _, err := os.Stat(dcy); os.IsNotExist(err) {
+		return fmt.Errorf("no such file: %s", flagConfig)
+	}
+
+	data, err := ioutil.ReadFile(dcy)
+	if err != nil {
+		return err
+	}
+
+	m, err := manifest.Load(data)
+	if err != nil {
+		return err
+	}
+
+	s := make(chan string)
+
+	go func() {
+		for l := range s {
+			log(l)
```

Comment:
Might not be a big deal overall, but we probably want to defer close the channel `s` so this loop will stop nicely.